File: week_6/sentimental-readability/readability.py
from cs50 import get_string


def count_words(text):
    """Calculates the number of words in a text by counting the number of spaces and
    adding 1."""
    count = 0

    for character in text:
        if character == " ":
            count += 1

    return count + 1


def count_letters(text):
    """Counts the number of letters in a text by counting alpha characters [a-zA-Z]."""
    count = 0

    for character in text:
        if character.isalpha():
            count += 1

    return count


def count_sentences(text):
    """Counts the number of sentences by counting the final characters [.!?]."""
    count = 0

    for character in text:
        if character in ".?!":
            count += 1

    return count


text = get_string("Text: ")

# Count the letters, words and sentences in the input text and calculate the averages
# per 100 words.
word_count = count_words(text)
letter_avg = 100.0 * count_letters(text) / word_count
sentence_avg = 100.0 * count_sentences(text) / word_count

# Calclulate the Coleman-Liau index.
index = 0.0588 * letter_avg - 0.296 * sentence_avg - 15.8

if index < 1:
    print("Before Grade 1")
elif index > 16:
    print("Grade 16+")
else:
    print(f"Grade {int(round(index))}")

File: week_2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int count_words(string text);
int count_letters(string text);
int count_sentences(string text);

int main(void)
{
    string text = get_string("Text: ");

    // Count the letters, words and sentences in the input text.
    int letter_count = count_letters(text);
    int word_count = count_words(text);
    int sentence_count = count_sentences(text);

    // Calclulate the Coleman-Liau index.
    float index = 0.0588 * ((float)letter_count / word_count * 100) -
                  0.296 * ((float)sentence_count / word_count * 100) - 15.8;

    if (index < 1)
    {
        // In case the index is before grade 1.
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        // In case the index is beyond grade 16.
        printf("Grade 16+\n");
    }
    else
    {
        // If not, print the rounded grade.
        printf("Grade %i\n", (int)round(index));
    }
}

int count_letters(string text)
{
    // Count the number of letters, by counting [a-z] and [A-Z] characters.
    int count = 0;

    for (int i = 0; i < strlen(text); i++)
    {
        if (isupper(text[i]) || islower(text[i]))
        {
            count++;
        }
    }

    return count;
}

int count_words(string text)
{
    // Count the number of words, by counting the spaces between them and adding one.
    int count = 0;

    for (int i = 0; i < strlen(text); i++)
    {
        if (text[i] == ' ')
        {
            count++;
        }
    }

    return ++count;
}

int count_sentences(string text)
{
    // Count the number of sentences, by counting the .?!
    int count = 0;

    for (int i = 0; i < strlen(text); i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count++;
        }
    }

    return count;
}